### ISSUE TITLE
TypeScript doesn't pick the correct overloading when mapping a ConcurrentFutureInstance (v12.0.0)

### ISSUE THREAD

**Issue by gcanti** [fluture-js/Fluture#400]
Repro

```ts
import * as F from 'fluture'

declare const x: F.FutureInstance<Error, number>
declare const y: F.ConcurrentFutureInstance<Error, number>

declare function double(n: number): number

F.map(double)(x) // ok

// v-- here is still picking the `FutureInstance` overloading
F.map(double)(y) // Argument of type 'ConcurrentFutureInstance<Error, number>' is not assignable to parameter of type 'FutureInstance<unknown, number>'
```

Possible solution (`index.d.ts`)

```ts
/** Map over the resolution value of the given Future / ConcurrentFuture. See https://github.com/fluture-js/Fluture#map */
export function map<RA, RB>(mapper: (value: RA) => RB): {
  <L>(source: FutureInstance<L, RA>): FutureInstance<L, RB>
  <L>(source: ConcurrentFutureInstance<L, RA>): ConcurrentFutureInstance<L, RB>
}
```

instead of

```ts
/** Map over the resolution value of the given Future. See https://github.com/fluture-js/Fluture#map */
export function map<RA, RB>(mapper: (value: RA) => RB): <L>(source: FutureInstance<L, RA>) => FutureInstance<L, RB>

/** Map over the resolution value of the given ConcurrentFuture. See https://github.com/fluture-js/Fluture#map */
export function map<RA, RB>(mapper: (value: RA) => RB): <L>(source: ConcurrentFutureInstance<L, RA>) => ConcurrentFutureInstance<L, RB>
```

**Comment by Avaq** [comment]
Thank you for the report! I imagine the same issue also affects other algebra's that use overloading in TypeScript, such as `chain`, `ap`, etc.

**Comment by gcanti** [comment]
> I imagine the same issue also affects other algebra's that use overloading in TypeScript

Actually no, they are fine, that's because TypeScript can use the first parameter (of `alt` / `ap`) to disambiguate

```ts
/** Logical or for Futures. See https://github.com/fluture-js/Fluture#alt */
export function alt<L, R>(left: FutureInstance<L, R>): (right: FutureInstance<L, R>) => FutureInstance<L, R>

/** Race two ConcurrentFutures. See https://github.com/fluture-js/Fluture#alt */
export function alt<L, R>(left: ConcurrentFutureInstance<L, R>): (right: ConcurrentFutureInstance<L, R>) => ConcurrentFutureInstance<L, R>

/** Apply the function in the left Future to the value in the right Future. See https://github.com/fluture-js/Fluture#ap */
export function ap<L, RA>(value: FutureInstance<L, RA>): <RB>(apply: FutureInstance<L, (value: RA) => RB>) => FutureInstance<L, RB>

/** Apply the function in the left ConcurrentFuture to the value in the right ConcurrentFuture. See https://github.com/fluture-js/Fluture#ap */
export function ap<L, RA>(value: ConcurrentFutureInstance<L, RA>): <RB>(apply: ConcurrentFutureInstance<L, (value: RA) => RB>) => ConcurrentFutureInstance<L, RB>
```

while in `map`

```ts
/** Map over the resolution value of the given Future. See https://github.com/fluture-js/Fluture#map */
export function map<RA, RB>(mapper: (value: RA) => RB): <L>(source: FutureInstance<L, RA>) => FutureInstance<L, RB>

/** Map over the resolution value of the given ConcurrentFuture. See https://github.com/fluture-js/Fluture#map */
export function map<RA, RB>(mapper: (value: RA) => RB): <L>(source: ConcurrentFutureInstance<L, RA>) => ConcurrentFutureInstance<L, RB>
```

can't do the same.

**Comment by Avaq** [comment]
Released as `12.0.1`.
